### ISSUE TITLE
Model Graph with Custom Training Loop

### ISSUE THREAD

**Issue by AnimeshSinha1309** [tensorflow/tensorboard#5264]
### Diagnostics

<details>
<summary>Diagnostics output</summary>

``````
--- check: autoidentify
INFO: diagnose_tensorboard.py version e43767ef2b648d0d5d57c00f38ccbd38390e38da

--- check: general
INFO: sys.version_info: sys.version_info(major=3, minor=9, micro=6, releaselevel='final', serial=0)
INFO: os.name: posix
INFO: os.uname(): posix.uname_result(sysname='Linux', nodename='animesh-laptop', release='5.11.0-27-generic', version='#29~20.04.1-Ubuntu SMP Wed Aug 11 15:58:17 UTC 2021', machine='x86_64')
INFO: sys.getwindowsversion(): N/A

--- check: package_management
INFO: has conda-meta: True
INFO: $VIRTUAL_ENV: None

--- check: installed_packages
INFO: installed: tensorboard==2.6.0
INFO: installed: tensorflow==2.6.0
INFO: installed: tensorflow-estimator==2.6.0
INFO: installed: tensorboard-data-server==0.6.1

--- check: tensorboard_python_version
INFO: tensorboard.version.VERSION: '2.6.0'

--- check: tensorflow_python_version
2021-08-23 10:06:37.151901: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-08-23 10:06:37.151935: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
INFO: tensorflow.__version__: '2.6.0'
INFO: tensorflow.__git_version__: 'v2.6.0-rc2-32-g919f693420e'

--- check: tensorboard_data_server_version
INFO: data server binary: '/home/animesh/Software/anaconda/envs/mol-generator/lib/python3.9/site-packages/tensorboard_data_server/bin/server'
INFO: data server binary version: b'rustboard 0.6.1'

--- check: tensorboard_binary_path
INFO: which tensorboard: b'/home/animesh/Software/anaconda/envs/mol-generator/bin/tensorboard\n'

--- check: addrinfos
socket.has_ipv6 = True
socket.AF_UNSPEC = <AddressFamily.AF_UNSPEC: 0>
socket.SOCK_STREAM = <SocketKind.SOCK_STREAM: 1>
socket.AI_ADDRCONFIG = <AddressInfo.AI_ADDRCONFIG: 32>
socket.AI_PASSIVE = <AddressInfo.AI_PASSIVE: 1>
Loopback flags: <AddressInfo.AI_ADDRCONFIG: 32>
Loopback infos: [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::1', 0, 0, 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 0))]
Wildcard flags: <AddressInfo.AI_PASSIVE: 1>
Wildcard infos: [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('0.0.0.0', 0)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::', 0, 0, 0))]

--- check: readable_fqdn
INFO: socket.getfqdn(): 'animesh-laptop'

--- check: stat_tensorboardinfo
INFO: directory: /tmp/.tensorboard-info
INFO: os.stat(...): os.stat_result(st_mode=16895, st_ino=31195481, st_dev=2056, st_nlink=2, st_uid=1000, st_gid=1000, st_size=4096, st_atime=1629686547, st_mtime=1629693384, st_ctime=1629693384)
INFO: mode: 0o40777

--- check: source_trees_without_genfiles
INFO: tensorboard_roots (1): ['/home/animesh/Software/anaconda/envs/mol-generator/lib/python3.9/site-packages']; bad_roots (0): []

--- check: full_pip_freeze
INFO: pip freeze --all:
absl-py==0.13.0
argon2-cffi==20.1.0
astunparse==1.6.3
async-generator==1.10
attrs==21.2.0
backcall==0.2.0
bleach==4.0.0
Bottleneck @ file:///tmp/build/80754af9/bottleneck_1607575130224/work
cachetools==4.2.2
certifi==2021.5.30
charset-normalizer==2.0.4
clang==5.0
click==8.0.1
configparser==5.0.2
cycler==0.10.0
debugpy==1.4.1
decorator==5.0.9
defusedxml==0.7.1
dgl==0.7.0
docker-pycreds==0.4.0
entrypoints==0.3
flatbuffers==1.12
fonttools==4.25.0
gast==0.4.0
gitdb==4.0.7
GitPython==3.1.18
google-auth==1.35.0
google-auth-oauthlib==0.4.5
google-pasta==0.2.0
googledrivedownloader==0.4
greenlet @ file:///tmp/build/80754af9/greenlet_1628888132713/work
grpcio==1.39.0
h5py==3.1.0
idna==3.2
ipykernel==6.1.0
ipython==7.26.0
ipython-genutils==0.2.0
ipywidgets==7.6.3
isodate==0.6.0
jedi==0.18.0
Jinja2==3.0.1
joblib==1.0.1
jsonschema==3.2.0
jupyter==1.0.0
jupyter-client==6.1.12
jupyter-console==6.4.0
jupyter-core==4.7.1
jupyterlab-pygments==0.1.2
jupyterlab-widgets==1.0.0
keras==2.6.0
Keras-Preprocessing==1.1.2
kiwisolver @ file:///tmp/build/80754af9/kiwisolver_1612281846643/work
Markdown==3.3.4
MarkupSafe==2.0.1
matplotlib @ file:///tmp/build/80754af9/matplotlib-suite_1628003481308/work
matplotlib-inline==0.1.2
mistune==0.8.4
mkl-fft==1.3.0
mkl-random @ file:///tmp/build/80754af9/mkl_random_1626186066731/work
mkl-service==2.4.0
munkres==1.1.4
nbclient==0.5.3
nbconvert==6.1.0
nbformat==5.1.3
nest-asyncio==1.5.1
networkx @ file:///tmp/build/80754af9/networkx_1627459939258/work
notebook==6.4.3
numexpr @ file:///tmp/build/80754af9/numexpr_1618856529730/work
numpy @ file:///tmp/build/80754af9/numpy_and_numpy_base_1626264704411/work
oauthlib==3.1.1
olefile==0.46
opt-einsum==3.3.0
packaging==21.0
pandas==1.3.1
pandocfilters==1.4.3
parso==0.8.2
pathtools==0.1.2
patsy==0.5.1
pexpect==4.8.0
pickleshare==0.7.5
Pillow @ file:///tmp/build/80754af9/pillow_1625655408865/work
pip==21.2.3
plotly @ file:///tmp/build/80754af9/plotly_1626182254919/work
prometheus-client==0.11.0
promise==2.3
prompt-toolkit==3.0.19
protobuf==3.17.3
psutil==5.8.0
ptyprocess==0.7.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycairo==1.19.1
Pygments==2.10.0
pyparsing @ file:///home/linux1/recipes/ci/pyparsing_1610983426697/work
pyrsistent==0.18.0
python-dateutil @ file:///tmp/build/80754af9/python-dateutil_1626374649649/work
python-louvain==0.15
pytz @ file:///tmp/build/80754af9/pytz_1612215392582/work
PyYAML==5.4.1
pyzmq==22.2.1
qtconsole==5.1.1
QtPy==1.9.0
rdflib==6.0.0
reportlab==3.5.67
requests==2.26.0
requests-oauthlib==1.3.0
rsa==4.7.2
scikit-learn==0.24.2
scipy @ file:///tmp/build/80754af9/scipy_1618855955483/work
seaborn @ file:///home/conda/feedstock_root/build_artifacts/seaborn-split_1611834504644/work
Send2Trash==1.8.0
sentry-sdk==1.3.1
setuptools==52.0.0.post20210125
shortuuid==1.0.1
six @ file:///tmp/build/80754af9/six_1605205306277/work
smmap==4.0.0
SQLAlchemy @ file:///tmp/build/80754af9/sqlalchemy_1626947882125/work
statsmodels @ file:///tmp/build/80754af9/statsmodels_1614022848006/work
subprocess32==3.5.4
tenacity @ file:///tmp/build/80754af9/tenacity_1626248292117/work
tensorboard==2.6.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.0
tensorflow==2.6.0
tensorflow-estimator==2.6.0
termcolor==1.1.0
terminado==0.11.0
testpath==0.5.0
threadpoolctl==2.2.0
torch==1.9.0
torch-cluster==1.5.9
torch-geometric==1.7.2
torch-scatter==2.0.8
torch-sparse==0.6.11
torch-spline-conv==1.2.1
tornado @ file:///tmp/build/80754af9/tornado_1606942317143/work
tqdm==4.62.1
traitlets==5.0.5
typing-extensions==3.7.4.3
urllib3==1.26.6
wandb==0.12.0
wcwidth==0.2.5
webencodings==0.5.1
Werkzeug==2.0.1
wheel==0.37.0
widgetsnbextension==3.5.1
wrapt==1.12.1

``````
</details>

## Issue description

I have a custom Keras model being trained with a custom training loop. I need an easy way to visualize the model graph (I can already see the scalars and the image logging). But the model graph never gets uploaded with the custom training loop. I am using TensorFlow 2.6. The complete code is attached in the following section.

I have tried the following idea, which is the only answer found anywhere else on the internet, on 2 questions on stack overflow and one other website. This DOES NOT help. Nothing gets stored in the log directories using this callback at all.

```python
    tb_callback = tf.keras.callbacks.TensorBoard('logs/' + current_time)
    tb_callback.set_model(model)
```

Do I need to use `tf.summary.trace_export("model-graph")`? I am using regular custom keras modules which derive from `tf.keras.models.Model` and `tf.keras.layers.Layer`, I don't have any `@tf.functions` I need graphed.

## Associated Questions

The same question has been asked on TensorFlow, but the answer doesn't help me:
* https://stackoverflow.com/questions/60639731/tensorboard-for-custom-training-loop-in-tensorflow-2
* https://stackoverflow.com/questions/61172053/tensorboard-graph-with-custom-training-loop-does-not-include-my-model/61173028#61173028

**Comment by AnimeshSinha1309** [comment]
I tried using the trace function, no luck there:
```python
tf.summary.trace_on(graph=True, profiler=False)
tf.summary.trace_export(name="model_trace", step=999)
```

I tried calling the TensorBoard Callback, doesn't work:
```python
tb_callback = tf.keras.callbacks.TensorBoard('logs/' + current_time)
callbacks = tf.keras.callbacks.CallbackList([tb_callback], add_history=True, model=model)
callbacks.on_train_begin()
```

The code with both those things ends up in this error, the Graph tab shows up, but the Graph wasn't uploaded.
![tensorboard-error-image](https://user-images.githubusercontent.com/9739218/130491681-38914be1-2c02-4445-bb78-b7c16c6b9572.png)

**Comment by stephanwlee** [comment]
Hi @AnimeshSinha1309, Keras callback is maintained by Keras team (https://github.com/keras-team/keras) so I may not be able to answer your question satisfactorily so consider moving your question there. Now, `tf.summary.trace_on` definitely is in realm of TensorBoard team so let's address that.

Is your code snippet an excerpt? When using trace API, you need to actually do something so a graph can be traced. That is,

```py
tf.summary.trace_on(graph=true)
# do something
my_custom_training_loop.step()
with writer.as_default():
  tf.summary.trace_export('model_trace', step=my_custom_training_loop_step_num)
```

**Comment by AnimeshSinha1309** [comment]
My full code looks more like this. The thing happening post trace was switched on to the end of tracing is not a `@tf.function`, but the model did get called, I will try moving it to a `@tf.function`.
```python
    tf.summary.trace_on(graph=True, profiler=False)
    
    with tqdm.trange(1, 1 + steps) as train_iterator:
        for step in train_iterator:
            idx_1 = np.random.randint(number_of_examples)
            idx_2 = np.random.choice([np.random.randint(number_of_examples), idx_1])

            with tf.GradientTape() as tape:
                mol_1, _label_1 = dataset[idx_1]
                mol_2, _label_2 = dataset[idx_2]
                logits = model(transformer(mol_1), transformer(mol_2))  # Logits for this mini-batch
                loss_value = loss_fn(tf.convert_to_tensor([idx_1 == idx_2]), logits)

            with train_summary_writer.as_default():
                if step == steps:
                    tf.summary.trace_export(name="model_trace", step=0)
                tf.summary.scalar('loss', loss_value, step=step)
```

**Comment by AnimeshSinha1309** [comment]
![image](https://user-images.githubusercontent.com/9739218/130580593-aa91ec83-7b89-4620-a0b8-71ee853c3c7d.png)

Having failed to make it work in TensorFlow, I switched over to PyTorch. But now it objects to me passing in Graphs instead of Tensors, the graphing still does not work.

**Comment by stephanwlee** [comment]
PyTorch is something that we definitely don't support. Please refer to their project for support.

As to the TF code you used, yes, you should be using `@tf.function` but your code snippet looks less than ideal. For one, you are turning on the trace once outside the step iteration yet you terminate it on each step. I don't think you can be terminating a trace you did not start. Anyways, if you want to still use Keras, please talk to their support and find out what they recommend for a custom Model.

**Comment by AnimeshSinha1309** [comment]
Accidentally terminating the trace gives an error, I have tried terminating it after 1 step and after all 999 steps, neither worked. The issue is not the use of Keras or PyTorch I believe, it's the problem of how to plot the graph of the network if the inputs to the network **is not a tensor or a list of tensors, it's a DGL graph object**. This is the reason why dummy input is not working in PyTorch, and that's why I believe Keras is unable to graph it. I removed the Graph stuff and all the codes work with regular objects. I will try to report this on the DGL repo I presume. Please close the issue if not not in the perview of the tensorboard team.

**Comment by AnimeshSinha1309** [comment]
I got the answer from the PyTorch Geometric people, the issue was caused because I was taking the graph as input to the model, where as I should have unpacked the tensors from the graph and fed them as input. So everything works now, thanks a lot.
